src/mlpack/methods/ann/layer/neural_turing_machine_impl.hpp

```diff
@@ -0,0 +1,249 @@
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_NTM_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_NTM_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "neural_turing_machine_impl.hpp"
+
+#include "../visitor/forward_visitor.hpp"
+#include "../visitor/forward_with_memory_visitor.hpp"
+#include "../visitor/backward_with_memory_visitor.hpp"
+#include "../visitor/backward_visitor.hpp"
+#include "../visitor/gradient_visitor.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template <typename InputDataType, typename OutputDataType>
+NeuralTuringMachine<InputDataType, OutputDataType>::NeuralTuringMachine(
+    const size_t inSize,
+    const size_t outSize,
+    const size_t numMem,
+    const size_t memSize,
+    const size_t shiftSize,
+    LayerTypes controller) :
+    inSize(inSize),
+    outSize(outSize),
+    numMem(numMem),
+    memSize(memSize),
+    shiftSize(shiftSize),
+    controller(controller),
+    deterministic(false)
+{
+  network.push_back(controller);
+
+  readMem = new ReadMemory<>(outSize, numMem, memSize, shiftSize);
+  writeMem = new WriteMemory<>(outSize, numMem, memSize, shiftSize);
+
+  network.push_back(readMem);
+  network.push_back(writeMem);
+
+  allZeros = arma::zeros<arma::mat>(memSize, 1);
+
+  allOnes = arma::ones<arma::mat>(numMem, memSize);
+
+  memoryHistory.push_back(std::move(arma::mat(allOnes.memptr(),
+      allOnes.n_rows, allOnes.n_cols, false, true)));
+
+  bMemoryHistory = memoryHistory.end();
+
+  lReads.push_back(std::move(arma::mat(allZeros.memptr(),
+      allZeros.n_rows, allZeros.n_cols, false, true)));
+
+  gReads = lReads.end();
+
+  forwardStep = 0;
+  backwardStep = 0;
+  gradientStep = 0;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void NeuralTuringMachine<InputDataType, OutputDataType>::Forward(
+    arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  // Create input to the controller.
+  arma::vec newInput = arma::join_vert(input, lReads.back());
+
+  // Forward pass through the controller.
+  boost::apply_visitor(ForwardVisitor(std::move(newInput), std::move(
+      boost::apply_visitor(outputParameterVisitor, controller))),
+      controller);
+
+  // Get controller output.
+  arma::mat& controllerOutput = boost::apply_visitor(outputParameterVisitor,
+      controller);
+
+  // Acess to current memory.
+  arma::mat& cMemory = memoryHistory.back();
+
+  // Pass the controller output through read memory layer.
+  boost::apply_visitor(ForwardWithMemoryVisitor(std::move(controllerOutput),
+      std::move(cMemory),
+      std::move(boost::apply_visitor(outputParameterVisitor, readMem))),
+      readMem);
+  lReads.push_back(boost::apply_visitor(outputParameterVisitor, readMem));
```

Comment:
ahh man. I totally agree. This is a better design. I feel like an idiot now :) It would add some more iterators in ReadMemory and WriteMemory but it would eliminate the use of ForwardWithMemory, which I always thought as a bad design but didn't think of any solution at that point. I totally missed this point that ReadMemory and WriteMemory are not made for FFN and RNN so we can create non-classical constructors for them.